src/libraries/System.Text.Json/gen/src/JsonSourceGenerator.cs

```diff
@@ -0,0 +1,117 @@
+
+        /// <summary>
+        /// Helper for unit tests.
+        /// TODO: refactor this after adopting emitter/parser pattern.
+        /// </summary>
+        public Dictionary<string, Type>? SerializableTypes => _helper.GetSerializableTypes();
+
+        /// <summary>
+        /// Registers a syntax resolver to receive compilation units.
+        /// </summary>
+        /// <param name="context"></param>
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            context.RegisterForSyntaxNotifications(() => new JsonSerializableSyntaxReceiver());
+        }
+
+        /// <summary>
+        /// Generates source code to optimize serialization and deserialization with JsonSerializer.
+        /// </summary>
+        /// <param name="executionContext"></param>
+        public void Execute(GeneratorExecutionContext executionContext)
+        {
+            Compilation compilation = executionContext.Compilation;
+
+            const string JsonSerializableAttributeName = "System.Text.Json.Serialization.JsonSerializableAttribute";
+            INamedTypeSymbol jsonSerializableAttribute = compilation.GetTypeByMetadataName(JsonSerializableAttributeName);
+            if (jsonSerializableAttribute == null)
+            {
+                return;
+            }
+
+            JsonSerializableSyntaxReceiver receiver = (JsonSerializableSyntaxReceiver)executionContext.SyntaxReceiver;
+            MetadataLoadContextInternal metadataLoadContext = new(compilation);
+
+            TypeExtensions.NullableOfTType = metadataLoadContext.Resolve(typeof(Nullable<>));
+
+            JsonSourceGeneratorHelper helper = new(executionContext, metadataLoadContext);
+            _helper = helper;
+
+            // Discover serializable types indicated by JsonSerializableAttribute.
+            foreach (CompilationUnitSyntax compilationUnit in receiver.CompilationUnits)
+            {
+                SemanticModel compilationSemanticModel = executionContext.Compilation.GetSemanticModel(compilationUnit.SyntaxTree);
+
+                foreach (AttributeListSyntax attributeListSyntax in compilationUnit.AttributeLists)
+                {
+                    AttributeSyntax attributeSyntax = attributeListSyntax.Attributes.Single();
+                    IMethodSymbol attributeSymbol = compilationSemanticModel.GetSymbolInfo(attributeSyntax).Symbol as IMethodSymbol;
+
+                    if (attributeSymbol == null || !jsonSerializableAttribute.Equals(attributeSymbol.ContainingType, SymbolEqualityComparer.Default))
+                    {
+                        // Not the right attribute.
+                        continue;
+                    }
+
+                    // Get JsonSerializableAttribute arguments.
+                    IEnumerable<SyntaxNode> attributeArguments = attributeSyntax.DescendantNodes().Where(node => node is AttributeArgumentSyntax);
+
+                    int argumentCount = attributeArguments.Count();
```

Comment:
Calling Count() here, then First() below, and then ElementAt() below are each going to enumerate this query again.  Is that necessary?

Reply:
I've cleaned this up, thanks